### DISCUSSION TITLE
assets:precompile task not detected

### DISCUSSION THREAD

**Discussion by surfcode** [ElMassimo/vite_ruby#366] [General]
I have been using Heroku which works with Vite Ruby out of the box. I tried out [Hatchbox](https://hatchbox.io/) the other day and noticed that it skips the precompilation step when deploying:

`-----> [Skip] assets:precompile task not detected`

While on Heroku it does detect the task:

`-----> Detecting rake tasks
-----> Preparing app for Rails asset pipeline
	   Running: rake assets:precompile`

At first I thought there was a bug as Chris Oliver, the creator of Hatchbox [states that the default deploy does in fact run the needed precompile task](https://hatchbox.relationkit.io/articles/55-what-is-the-default-rails-deploy-script). So I opened a ticket but got  the following answer:

_Hatchbox does precompile the assets but not if it is skipped due to a change in the basic Rails setup. It is possible that Vite or something else is causing Hatchbox to skip that step. If you change that Hatchbox will precompile them for you._

Now I am wondering what needs to be adjusted exactly. I looked up the source code here https://github.com/ElMassimo/vite_ruby/blob/main/vite_ruby/lib/tasks/vite.rake but I'm unsure whether there's anything being adjusted in the Rails setup itself as the answer states.

Curious to know what needs to be changed for the assets:precompile task to be executed automatically just like on Heroku.

Best regards

**Comment by ElMassimo** [comment]
Hi there!

Comparing the [Hatchbox deploy script](https://hatchbox.relationkit.io/articles/55-what-is-the-default-rails-deploy-script), and the output log you provided:

> assets:precompile task not detected

suggests that Hatchbox is doing something else to detect the task.

I'd suggest following up with Hatchbox to understand how they are checking if the task is defined.

Perhaps they are checking for specific gems, like `sprockets`, `propshaft`, or `webpacker`, in which case you could ask them to provide a fix or workaround, or use a custom deploy script.

**Comment by surfcode** [comment]
Hey @ElMassimo!

You are absolutely right. I just received the following message from Chris Oliver:

> The "rails -T" command does not show an assets:precompile command, so that's why Hatchbox does not run it. The command isn't available, so you'd just get an error saying that command is missing.
> 
> I see you've commented out sprockets which provides the assets:precompile step. The easiest fix would be to add it back in, or define your own assets:precompile rake task that does nothing.
> 
> As per the docs, Vite will automatically run its commands after assets:precompile, so you'll need that command or you will need to define a custom build script in Hatchbox to run the vite commands directly. 

I'm new to Ruby on Rails and thought `sprockets` would be unnecessary overhead in conjunction with Vite Ruby and indeed I have the following in my application.rb:

`#require "sprockets/railtie"`

 Can it be safely uncommented? That would apparently provide the immediate fix. Or maybe his other suggestion might be a more lean way of triggering the precompile step. I guess I need to  put a file under the lib/tasks folder in my rails app in that case.

**Reply by ElMassimo** [reply]
Cool, now it makes sense that it skips the task.

The reason is that these "fallback" rake tasks that Vite Ruby defines simply don't have a description, so they are available but don't show up when doing `rails -T` (which calls `rake -T`, which by default skips tasks without a description).

Since it would be beneficial for these tasks to show up in the list, I've [pushed a commit](https://github.com/ElMassimo/vite_ruby/commit/66af6eb0268e7a989562feb38da74072817f0d49) to add comments to these tasks.

Released in `vite_ruby-3.3.2`, please upgrade and let me know if that does the trick 😃

**Reply by surfcode** [reply]
Wow. Fantastic! Thank you for the clear explanation and the immediate fix! I'll let you know in a few minutes.

**Reply by surfcode** [reply]
```
-----> /home/deploy/.asdf/bin/asdf exec bundle exec rake assets:precompile
...
Building with Vite ⚡️
vite v4.3.5 building for production...
...
Build with Vite complete: /home/deploy/red-wildflower-595/releases/20230509193137/public/vite
```

It works like a charm. Awesome!
